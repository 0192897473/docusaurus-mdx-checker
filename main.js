#!/usr/bin/env node

import fs from "node:fs/promises";
import process from "node:process";
import path from "node:path";
import { globby } from "globby";
import { compile } from "@mdx-js/mdx";
import remarkFrontmatter from "remark-frontmatter";
import remarkDirectives from "remark-directive";
import remarkGFM from "remark-gfm";
import remarkComment from "@slorber/remark-comment";

////////////////////////////////////////
// Params

const verbose = false;

const excludeVersionedDocs = true;

const cwd = "/Users/sebastienlorber/Desktop/projects/jest";

const include = ["**/*.{md,mdx}", "../docs/**/*.{md,mdx}"];

const exclude = [
  "packages",
  "examples",
  "node_modules",
  //
  "CHANGELOG.md",
  "CODE_OF_CONDUCT.md",
  "CONTRIBUTING.md",
  "README.md",
  "website/README.md",
  "benchmarks/test-file-overhead/README.md",
  //
  excludeVersionedDocs ? "**/versioned_docs" : null,
].filter(Boolean);

const format = "mdx";

const remarkPlugins = [
  remarkFrontmatter,
  remarkDirectives,
  remarkGFM,
  remarkComment,
];

////////////////////////////////////////
// Script

const allRelativeFilePaths = await globby(include, {
  cwd,
  ignore: exclude,
  // gitignore: true, // Does not work well with relative paths like ../docs
});

if (allRelativeFilePaths.length === 0) {
  console.error("Couldn't find any file to compile!");
  process.exit(1);
}

console.log(
  "Found " + allRelativeFilePaths.length + " files to compile with MDX"
);

if (verbose) {
  console.log(
    "List of files:\n",
    JSON.stringify(allRelativeFilePaths, null, 2)
  );
}

const allResults = await Promise.all(allRelativeFilePaths.map(processFilePath));

const allErrors = allResults.filter((r) => r.status === "error");
const allSuccess = allResults.filter((r) => r.status === "success");

console.log(`Errors: ${allErrors.length}`);
console.log(`Success: ${allSuccess.length}`);

if (allErrors.length === 0) {
  console.log("All MDX files compile successfully!");
  process.exit(0);
}

const outputSeparator = "\n---\n";
const outputs = allErrors.map((error) => {
  return error.errorMessage;
});

console.log("Some MDX files couldn't compile successfully!");
console.log(outputSeparator + outputs.join(outputSeparator) + outputSeparator);

process.exit(1);

////////////////////////////////////////
// Functions

async function processFilePath(relativeFilePath) {
  const filePath = path.resolve(cwd, relativeFilePath);
  try {
    // console.log("filePath", filePath);
    const fileContent = await fs.readFile(filePath);
    // console.log("fileContent", fileContent);
    const result = await compile(fileContent, { format, remarkPlugins });
    // TODO generate warnings for compat options here?
    return { relativeFilePath, status: "success", result };
  } catch (error) {
    const errorMessage = `Error while compiling file ${relativeFilePath}:\n${formatMDXErrorMessage(
      error
    )}`;
    return { relativeFilePath, status: "error", error, errorMessage };
  }
}

function formatMDXErrorMessage(error) {
  let message = `Details: ${error.message}`;

  if (error.line || error.column) {
    let lineColumn = "";
    if (error.line) {
      lineColumn = lineColumn + "Line=" + error.line + " ";
    }
    if (error.column) {
      lineColumn = lineColumn + "Column=" + error.column;
    }
    message = message + "\n" + lineColumn;
  }

  return message;
}
