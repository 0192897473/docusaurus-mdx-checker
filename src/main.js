#!/usr/bin/env node

import fs from "node:fs/promises";
import path from "node:path";
import { globby } from "globby";
import { compile } from "@mdx-js/mdx";
import chalk from "chalk";

import { preprocess } from "./compat.js";
import {
  DefaultRemarkPlugins,
  DefaultRehypePlugins,
  DefaultInclude,
  DefaultExclude,
} from "./constants.js";

export {
  DefaultRemarkPlugins,
  DefaultRehypePlugins,
  DefaultInclude,
  DefaultExclude,
} from "./constants";

const SuccessPrefix = chalk.green("[SUCCESS]");
const ErrorPrefix = chalk.red("[ERROR]");

////////////////////////////////////////
// Script

export default async function main({
  cwd = undefined,
  include = DefaultInclude,
  exclude = DefaultExclude,
  remarkPlugins = DefaultRemarkPlugins,
  rehypePlugins = DefaultRehypePlugins,
  format = "mdx",
  verbose = false,
}) {
  const allRelativeFilePaths = await globby(include, {
    cwd,
    ignore: exclude,
    // gitignore: true, // Does not work well with relative paths like ../docs
  });
  allRelativeFilePaths.sort();

  if (allRelativeFilePaths.length === 0) {
    throw new Error(`${ErrorPrefix} Couldn't find any file to compile!`);
  }

  console.log(
    "Found " + allRelativeFilePaths.length + " files to compile with MDX"
  );

  if (verbose) {
    console.log(
      "List of files:\n",
      JSON.stringify(allRelativeFilePaths, null, 2)
    );
  }

  const allResults = await Promise.all(
    allRelativeFilePaths.map(processFilePath)
  );

  const allErrors = allResults.filter((r) => r.status === "error");
  const allSuccess = allResults.filter((r) => r.status === "success");

  console.log(`Errors: ${allErrors.length}`);
  console.log(`Success: ${allSuccess.length}`);

  if (allErrors.length > 0) {
    const outputSeparator = `\n${chalk.yellow("---")}\n`;
    throw new Error(
      `${ErrorPrefix} Some MDX files couldn't compile successfully!
${outputSeparator}${allErrors
        .map((error) => error.errorMessage)
        .join(outputSeparator)}${outputSeparator}`
    );
  } else {
    return `${SuccessPrefix} All ${allRelativeFilePaths.length} MDX files compiled successfully!`;
  }

  ////////////////////////////////////////
  // Functions

  async function processFilePath(relativeFilePath) {
    const filePath = path.resolve(cwd, relativeFilePath);
    try {
      // console.log("filePath", filePath);
      const fileContent = await fs.readFile(filePath, "utf8");
      // console.log("fileContent", fileContent);
      const contentPreprocessed = preprocess(fileContent);
      const result = await compile(contentPreprocessed, {
        format:
          format === "detect"
            ? filePath.endsWith(".md")
              ? "md"
              : "mdx"
            : "mdx",
        remarkPlugins,
        rehypePlugins,
      });
      // TODO generate warnings for compat options here?
      return { relativeFilePath, status: "success", result };
    } catch (error) {
      const errorMessage = `Error while compiling file ${relativeFilePath}:\n${formatMDXErrorMessage(
        error
      )}`;
      return { relativeFilePath, status: "error", error, errorMessage };
    }
  }

  function formatMDXErrorMessage(error) {
    let message = `Details: ${error.message}`;

    if (error.line || error.column) {
      let lineColumn = "";
      if (error.line) {
        lineColumn = lineColumn + "Line=" + error.line + " ";
      }
      if (error.column) {
        lineColumn = lineColumn + "Column=" + error.column;
      }
      message = message + "\n" + lineColumn;
    }

    return message;
  }
}
