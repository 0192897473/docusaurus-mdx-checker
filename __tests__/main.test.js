import { describe, expect, test } from "vitest";
import main from "../src/main";
import {
  DefaultRemarkPlugins,
  VersionedDocsExclusion,
} from "../src/constants.js";
import remarkMath from "remark-math";

const FixtureSitesPath = "./__tests__/__fixtures__/sites";

const SiteFixtures = {
  v2: {
    docusaurus: `${FixtureSitesPath}/docusaurus-v2`,
    jest: `${FixtureSitesPath}/jest-v2`,
    reactnative: `${FixtureSitesPath}/react-native-v2`,
  },
  v3: {
    docusaurus: `${FixtureSitesPath}/docusaurus-v3`,
    jest: `${FixtureSitesPath}/jest-v3`,
    reactnative: `${FixtureSitesPath}/react-native-v3`,
  },
};

const exclude = [
  "packages",
  "examples",
  "node_modules",
  //
  "CHANGELOG.md",
  "CODE_OF_CONDUCT.md",
  "CONTRIBUTING.md",
  "README.md",
  "website/README.md",
  "benchmarks/test-file-overhead/README.md",
  //
  // excludeVersionedDocs ? "**/versioned_docs" : null,
].filter(Boolean);

const OnlyDocsInclude = "docs/**/*.{md,mdx}";

describe("Docusaurus v2", () => {
  async function testSite(options) {
    return main({
      cwd: SiteFixtures.v2.docusaurus,
      format: "mdx",
      remarkPlugins: [...DefaultRemarkPlugins, remarkMath],
      // verbose: true,
      ...options,
    });
  }

  test("does not compile", async () => {
    await expect(testSite({})).rejects.toThrowErrorMatchingSnapshot();
  });

  test("does not compile - only docs", async () => {
    await expect(
      testSite({
        include: [OnlyDocsInclude],
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test("does not compile - exclude versioned docs", async () => {
    await expect(
      testSite({
        exclude: [VersionedDocsExclusion],
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe("Docusaurus v3", () => {
  async function testSite(options) {
    return main({
      cwd: SiteFixtures.v3.docusaurus,
      format: "detect",
      remarkPlugins: [...DefaultRemarkPlugins, remarkMath],
      // verbose: true,
      ...options,
    });
  }

  test("compiles", async () => {
    const result = await testSite({});

    expect(result).toMatchInlineSnapshot(
      '"[32m[SUCCESS][39m All 694 MDX files compiled successfully!"'
    );
  });

  test("compiles - only docs", async () => {
    const result = await testSite({
      include: [OnlyDocsInclude],
    });

    expect(result).toMatchInlineSnapshot(
      '"[32m[SUCCESS][39m All 85 MDX files compiled successfully!"'
    );
  });

  test("compiles - exclude versioned docs", async () => {
    const result = await testSite({
      exclude: [VersionedDocsExclusion],
    });

    expect(result).toMatchInlineSnapshot(
      '"[32m[SUCCESS][39m All 204 MDX files compiled successfully!"'
    );
  });
});
