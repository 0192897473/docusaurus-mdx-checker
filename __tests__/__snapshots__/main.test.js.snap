// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Docusaurus v2 > does not compile - exclude versioned docs 1`] = `
"[31m[ERROR][39m Some MDX files couldn't compile successfully!

[33m---[39m
Error while compiling file docs/api/plugins/plugin-pwa.md:
Details: Unexpected character \`=\` (U+003D) before name, expected a character that can start a name, such as a letter, \`$\`, or \`_\`
Line=123 Column=48
[33m---[39m
Error while compiling file docs/deployment.mdx:
Details: Unexpected closing tag \`</details>\`, expected corresponding closing tag for \`<TabItem>\` (320:1-320:48)
Line=443 Column=1
[33m---[39m
Error while compiling file docs/introduction.md:
Details: Unexpected character \`/\` (U+002F) after self-closing slash, expected \`>\` to end the tag (note: JS comments in JSX tags are not supported in MDX)
Line=57 Column=6
[33m---[39m
"
`;

exports[`Docusaurus v2 > does not compile - only docs 1`] = `
"[31m[ERROR][39m Some MDX files couldn't compile successfully!

[33m---[39m
Error while compiling file docs/api/plugins/plugin-pwa.md:
Details: Unexpected character \`=\` (U+003D) before name, expected a character that can start a name, such as a letter, \`$\`, or \`_\`
Line=123 Column=48
[33m---[39m
Error while compiling file docs/deployment.mdx:
Details: Unexpected closing tag \`</details>\`, expected corresponding closing tag for \`<TabItem>\` (320:1-320:48)
Line=443 Column=1
[33m---[39m
Error while compiling file docs/introduction.md:
Details: Unexpected character \`/\` (U+002F) after self-closing slash, expected \`>\` to end the tag (note: JS comments in JSX tags are not supported in MDX)
Line=57 Column=6
[33m---[39m
"
`;

exports[`Docusaurus v2 > does not compile 1`] = `
"[31m[ERROR][39m Some MDX files couldn't compile successfully!

[33m---[39m
Error while compiling file docs/api/plugins/plugin-pwa.md:
Details: Unexpected character \`=\` (U+003D) before name, expected a character that can start a name, such as a letter, \`$\`, or \`_\`
Line=123 Column=48
[33m---[39m
Error while compiling file docs/deployment.mdx:
Details: Unexpected closing tag \`</details>\`, expected corresponding closing tag for \`<TabItem>\` (320:1-320:48)
Line=443 Column=1
[33m---[39m
Error while compiling file docs/introduction.md:
Details: Unexpected character \`/\` (U+002F) after self-closing slash, expected \`>\` to end the tag (note: JS comments in JSX tags are not supported in MDX)
Line=57 Column=6
[33m---[39m
Error while compiling file versioned_docs/version-2.0.0-beta.22/api/plugins/plugin-pwa.md:
Details: Unexpected character \`=\` (U+003D) before name, expected a character that can start a name, such as a letter, \`$\`, or \`_\`
Line=123 Column=48
[33m---[39m
Error while compiling file versioned_docs/version-2.0.0-beta.22/deployment.mdx:
Details: Unexpected closing tag \`</details>\`, expected corresponding closing tag for \`<TabItem>\` (320:1-320:48)
Line=443 Column=1
[33m---[39m
Error while compiling file versioned_docs/version-2.0.0-beta.22/introduction.md:
Details: Unexpected character \`/\` (U+002F) after self-closing slash, expected \`>\` to end the tag (note: JS comments in JSX tags are not supported in MDX)
Line=57 Column=6
[33m---[39m
Error while compiling file versioned_docs/version-2.0.0-rc.1/api/plugins/plugin-pwa.md:
Details: Unexpected character \`=\` (U+003D) before name, expected a character that can start a name, such as a letter, \`$\`, or \`_\`
Line=123 Column=48
[33m---[39m
Error while compiling file versioned_docs/version-2.0.0-rc.1/deployment.mdx:
Details: Unexpected closing tag \`</details>\`, expected corresponding closing tag for \`<TabItem>\` (320:1-320:48)
Line=443 Column=1
[33m---[39m
Error while compiling file versioned_docs/version-2.0.0-rc.1/introduction.md:
Details: Unexpected character \`/\` (U+002F) after self-closing slash, expected \`>\` to end the tag (note: JS comments in JSX tags are not supported in MDX)
Line=57 Column=6
[33m---[39m
"
`;
